### ISSUE TITLE
[BUG] Datasets information are completely wrong, because different datasets are considered the same

### ISSUE THREAD

**Issue by fzyzcjy** [mlflow/mlflow#11560]
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

- Client: latest
- Tracking server: 1.x.y

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **Python version**:
- **yarn version, if running the dev UI**:

### Describe the problem

Hi thanks for the library! However,

### Tracking information

```shell
REPLACE_ME
```

### Code to reproduce issue

```
REPLACE_ME
```

### Stack trace

```
REPLACE_ME
```

### Other info / logs

```
REPLACE_ME
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by B-Step62** [comment]
@fzyzcjy Thank you for bringing the issue to our attention! Yes the dataset digest is computed with the first 10k rows for now. We need some limit to avoid hanging when calculating a digest for a huge dataset, but I think it's possible to make it configurable via a constructor argument. Will it be helpful for your use case?

**Comment by fzyzcjy** [comment]
@B-Step62 Thanks for the reply. I am not sure whether it is the best solution, because I will never realize I need to change this argument unless I realize it is mlflow's bug (instead of my code's bug).

PS. I guess another solution is that, even if two datasets have the same digest, mlflow can store both of them (instead of only storing one).

**Comment by B-Step62** [comment]
@fzyzcjy The digest is designed to be an unique identifier for a dataset combined with a name, so we shouldn't create two different dataset for a same digest, similarly to Git commit hash.

**Comment by fzyzcjy** [comment]
I see, however the git commit hash is computed for the whole commit IIRC, thus it is really unique (up to collision), while mlflow is not. On the other hand, I do agree we cannot compute for the whole dataset, which is too slow...

**Comment by B-Step62** [comment]
Yes, and the best value for the number of rows to be used for digesting depends on the dataset. 10k doesn't work for this case, but if a dataset contains many number of columns, even lesser number may be ideal. I understand it can be confusing, but since there is no single number satisfies every use case, I would simply make it configurable.

**Comment by fzyzcjy** [comment]
By the way, what if we consider number of rows when computing the digest? That will fix this specific case at least.

**Comment by B-Step62** [comment]
Do you mean increasing the number when the dataset contains more rows? It slows down logging large datasets.

**Comment by B-Step62** [comment]
> By the way, what if we consider number of rows when computing the digest? That will fix this specific case at least.

That sounds like a sweat idea! Indeed the hash function itself considers the length of dataset ([code](https://github.com/mlflow/mlflow/blame/1ad9148fb828a2f7beaad8394967a0691344ff81/mlflow/data/digest_utils.py#L35)), it's just that the passed dataset is always truncated to 10k rows so it is no-op for HF dataset ([code](https://github.com/mlflow/mlflow/blob/1ad9148fb828a2f7beaad8394967a0691344ff81/mlflow/data/huggingface_dataset.py#L63-L68)). If we can make it to consider the original length of the dataset, that would be awesome:)

**Comment by robmcd** [comment]
On a related topic, I've noticed that dataset hash will be _different_ even if the only difference between two datasets is the pandas dataframe index. Is there a good reason to consider different index values as different datasets? Or should we perhaps just ignore the index `pandas.util.hash_pandas_object(df, index=False)`?

**Comment by lxhil1217** [comment]
> @fzyzcjy The digest is designed to be an unique identifier for a dataset combined with a name, so we shouldn't create two different dataset for a same digest, similarly to Git commit hash.

So the same name combined with the same digest will always result in the same dataset, right? Could you please point the implementation of this logic? Thanks!
